server/src/main/scala/org/http4s/server/websocket/WebSocketBuilder2.scala

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package server.websocket
+
+import cats.{Applicative, ~>}
+import cats.syntax.all._
+import fs2.{Pipe, Stream}
+import org.http4s.websocket.{
+  WebSocket,
+  WebSocketCombinedPipe,
+  WebSocketContext,
+  WebSocketFrame,
+  WebSocketSeparatePipe
+}
+import org.typelevel.vault.Key
+
+/** Build a response which will accept an HTTP websocket upgrade request and initiate a websocket connection using the
+  * supplied exchange to process and respond to websocket messages.
+  * @param headers Handshake response headers, such as such as:Sec-WebSocket-Protocol.
+  * @param onNonWebSocketRequest The status code to return to a client making a non-websocket HTTP request to this route.
+  *                              default: NotImplemented
+  * @param onHandshakeFailure The status code to return when failing to handle a websocket HTTP request to this route.
+  *                           default: BadRequest
+  */
+final case class WebSocketBuilder2[F[_]: Applicative](
```

Comment:
Making this a case class is how we got pinned in the first place.  Case class builders are bad, and not used most other places.  I'll PR a counterproposal.